flow/damage_context.cc

```diff
@@ -0,0 +1,139 @@
+  return comparator == e.comparator && paint_bounds == e.paint_bounds &&
+         std::equal(
+             mutators.begin(), mutators.end(), e.mutators.begin(),
+             e.mutators.end(),
+             [](const std::shared_ptr<Mutator>& m1,
+                const std::shared_ptr<Mutator>& m2) { return *m1 == *m2; }) &&
+         (layer.get() == e.layer.get() ||
+          comparator(layer.get(), e.layer.get()));
+}
+
+void DamageContext::InitFrame(const SkISize& tree_size,
+                              const FrameDescription* previous_frame) {
+  current_layer_tree_size_ = tree_size;
+  previous_frame_ = previous_frame;
+}
+
+void DamageContext::PushLayerEntry(const Layer* layer,
+                                   LayerComparator comparator,
+                                   const SkMatrix& matrix,
+                                   const PrerollContext& preroll_context,
+                                   size_t index) {
+  if (current_layer_tree_size_.isEmpty() || layer->paint_bounds().isEmpty()) {
+    return;
+  }
+
+  LayerEntry e;
+  e.layer = layer->shared_from_this();
+  e.comparator = comparator;
+
+  SkRect bounds = layer->paint_bounds();
+  bounds.intersect(preroll_context.cull_rect);
+  e.paint_bounds = matrix.mapRect(bounds);
+
+  for (auto i = preroll_context.mutators_stack.Begin();
+       i != preroll_context.mutators_stack.End(); ++i) {
+    auto type = (*i)->GetType();
+    // transforms are irrelevant because we compare paint_bounds in
+    // screen coordinates
+    if (type != MutatorType::transform) {
+      e.mutators.push_back(*i);
+    }
+  }
+  if (index == size_t(-1)) {
+    layer_entries_.push_back(std::move(e));
+  } else {
+    layer_entries_.insert(layer_entries_.begin() + index, std::move(e));
+  }
+}
+
+bool DamageContext::ApplyImageFilter(size_t from,
+                                     size_t count,
+                                     const SkImageFilter* filter,
+                                     const SkMatrix& matrix,
+                                     const SkRect& bounds) {
+  SkMatrix inverted;
+  if (!matrix.invert(&inverted)) {
+    return false;
+  }
+
+  for (size_t i = from; i < count; ++i) {
+    auto& entry = layer_entries_[i];
+    SkRect layer_bounds(entry.paint_bounds);
+    inverted.mapRect(layer_bounds);
+    if (layer_bounds.intersects(bounds)) {
+      // layer paint bounds after filter can not get bigger than union of
+      // original paint bounds and filter bounds
+      SkRect max(layer_bounds);
+      max.join(bounds);
+
+      layer_bounds = filter->computeFastBounds(layer_bounds);
+      layer_bounds.intersect(max);
+
+      matrix.mapRect(&layer_bounds);
+      entry.paint_bounds = layer_bounds;
+    }
+  }
+
+  return true;
+}
+
+DamageContext::DamageResult DamageContext::FinishFrame() {
+  DamageResult res;
+  LayerEntrySet entries;
+  entries.insert(layer_entries_.begin(), layer_entries_.end());
```

Comment:
Order matters. If we draw a rectangle and then a circle in one frame and in the next we've reversed their order and drawn the circle and then the rectangle (ignoring the fact that simply drawing a rectangle and a circle would get combined into a Picture, but generalize those concepts to 2 different engine layers that change order) - we need to repaint them. Using a set says that as long as they are still in the scene in the same device location we don't care when they get repainted. That is not true.

Reply:
Indeed. Thanks for pointing this out. I'll address this.

Reply:
https://github.com/knopp/engine/commit/2ae2bf35f6743dfbca924436731262108d4c0ce3 should take care of layers that are identical but painted in different order.